File: detector.py
import os
import cv2
import numpy as np

try:
    import ncnn
except ImportError:
    ncnn = None

VEHICLE_CLASSES = {2: 'car', 3: 'motorcycle', 5: 'bus', 7: 'truck'}


class YOLO26nDetector:
    def __init__(self, model_path, conf_thresh=0.5, target_size=320):
        self.conf_thresh = conf_thresh
        self.target_size = target_size
        self.vehicle_ids = set(VEHICLE_CLASSES.keys())
        self.net = None

        if ncnn is None:
            raise ImportError("ncnn not installed")

        param_path, bin_path = None, None
        if os.path.isdir(model_path):
            for f in os.listdir(model_path):
                if f.endswith('.param'):
                    param_path = os.path.join(model_path, f)
                elif f.endswith('.bin'):
                    bin_path = os.path.join(model_path, f)
        if param_path is None:
            param_path = f"{model_path}.param"
        if bin_path is None:
            bin_path = f"{model_path}.bin"

        self.net = ncnn.Net()
        self.net.opt.use_vulkan_compute = False
        self.net.load_param(param_path)
        self.net.load_model(bin_path)

    def detect(self, frame, enabled_classes=None):
        h, w = frame.shape[:2]
        ts = self.target_size

        scale = min(ts / h, ts / w)
        new_h, new_w = int(h * scale), int(w * scale)
        pad_h = (ts - new_h) // 2
        pad_w = (ts - new_w) // 2

        resized = cv2.resize(frame, (new_w, new_h))
        canvas = np.full((ts, ts, 3), 114, dtype=np.uint8)
        canvas[pad_h:pad_h + new_h, pad_w:pad_w + new_w] = resized

        in_mat = ncnn.Mat.from_pixels(canvas, 2, ts, ts)
        in_mat.substract_mean_normalize([], [1 / 255.0] * 3)

        ex = self.net.create_extractor()
        ex.input("in0", in_mat)
        _, out = ex.extract("out0")

        detections = self._postprocess(out, scale, pad_w, pad_h, w, h)
        class_ids = set(enabled_classes) if enabled_classes else self.vehicle_ids
        return [d for d in detections if d['class_id'] in class_ids]

    def _postprocess(self, out, scale, pad_w, pad_h, orig_w, orig_h):
        out_np = np.array(out)
        if out_np.ndim == 3:
            out_np = out_np[0]
        out_np = out_np.T

        results = []
        for i in range(out_np.shape[0]):
            vals = out_np[i]
            cx, cy, w, h = vals[:4]
            scores = vals[4:]
            cls_id = int(np.argmax(scores))
            conf = float(scores[cls_id])
            if conf < self.conf_thresh:
                continue

            x1 = int(max(0, min((cx - w / 2 - pad_w) / scale, orig_w)))
            y1 = int(max(0, min((cy - h / 2 - pad_h) / scale, orig_h)))
            x2 = int(max(0, min((cx + w / 2 - pad_w) / scale, orig_w)))
            y2 = int(max(0, min((cy + h / 2 - pad_h) / scale, orig_h)))

            if x2 <= x1 or y2 <= y1:
                continue

            results.append({
                'bbox': (x1, y1, x2, y2),
                'centroid': ((x1 + x2) // 2, (y1 + y2) // 2),
                'confidence': conf,
                'class_id': cls_id,
                'label': VEHICLE_CLASSES.get(cls_id, f'cls_{cls_id}'),
            })

        results.sort(key=lambda r: r['confidence'], reverse=True)
        keep = []
        for r in results:
            if not any(iou(r['bbox'], k['bbox']) > 0.45 for k in keep):
                keep.append(r)
        return keep


def iou(a, b):
    x1 = max(a[0], b[0])
    y1 = max(a[1], b[1])
    x2 = min(a[2], b[2])
    y2 = min(a[3], b[3])
    inter = max(0, x2 - x1) * max(0, y2 - y1)
    area_a = (a[2] - a[0]) * (a[3] - a[1])
    area_b = (b[2] - b[0]) * (b[3] - b[1])
    return inter / (area_a + area_b - inter + 1e-6)

File: rust_port/src/detector.rs
use crate::types::Detection;
use crate::ncnn_wrapper;
use opencv::prelude::*;

const COCO_CLASSES: &[&str] = &[
    "person","bicycle","car","motorcycle","airplane","bus","train","truck","boat",
    "traffic light","fire hydrant","stop sign","parking meter","bench","bird","cat",
    "dog","horse","sheep","cow","elephant","bear","zebra","giraffe","backpack",
    "umbrella","handbag","tie","suitcase","frisbee","skis","snowboard","sports ball",
    "kite","baseball bat","baseball glove","skateboard","surfboard","tennis racket",
    "bottle","wine glass","cup","fork","knife","spoon","bowl","banana","apple",
    "sandwich","orange","broccoli","carrot","hot dog","pizza","donut","cake","chair",
    "couch","potted plant","bed","dining table","toilet","tv","laptop","mouse",
    "remote","keyboard","cell phone","microwave","oven","toaster","sink","refrigerator",
    "book","clock","vase","scissors","teddy bear","hair drier","toothbrush",
];

pub struct Detector {
    net: ncnn_wrapper::NcnnNet,
    target_size: i32,
    conf_thresh: f32,
}

impl Detector {
    pub fn new(model_path: &str, target_size: u32) -> Result<Self, String> {
        let ts = target_size as i32;
        let model_dir = if model_path.ends_with("_ncnn_model") || model_path.contains("ncnn") {
            model_path.to_string()
        } else {
            format!("{}_ncnn_model", model_path.trim_end_matches(".onnx"))
        };

        let param_path = if model_path.ends_with(".param") {
            model_path.to_string()
        } else {
            format!("{}/model.ncnn.param", model_dir)
        };
        let bin_path = if model_path.ends_with(".bin") {
            model_path.to_string()
        } else {
            format!("{}/model.ncnn.bin", model_dir)
        };

        let param_path = resolve_path(&param_path);
        let bin_path = resolve_path(&bin_path);

        let net = ncnn_wrapper::NcnnNet::new().ok_or("ncnn net create failed")?;
        net.load_param(&param_path)?;
        net.load_model(&bin_path)?;

        Ok(Self { net, target_size: ts, conf_thresh: 0.5 })
    }

    pub fn detect(&mut self, mat: &opencv::core::Mat, enabled_classes: &[i32]) -> Vec<Detection> {
        let ts = self.target_size;
        let (fh, fw) = (mat.rows(), mat.cols());
        if fh <= 0 || fw <= 0 { return vec![]; }

        // Letterbox
        let scale = (ts as f32 / fw as f32).min(ts as f32 / fh as f32);
        let new_w = (fw as f32 * scale) as i32;
        let new_h = (fh as f32 * scale) as i32;
        let pad_x = (ts - new_w) / 2;
        let pad_y = (ts - new_h) / 2;

        // Build padded letterbox image using OpenCV (fast)
        let mut resized = opencv::core::Mat::default();
        if opencv::imgproc::resize(mat, &mut resized, opencv::core::Size::new(new_w, new_h), 0.0, 0.0, opencv::imgproc::INTER_LINEAR).is_err() {
            return vec![];
        }

        let mut letterbox = match opencv::core::Mat::new_rows_cols_with_default(ts, ts, opencv::core::CV_8UC3, opencv::core::Scalar::new(128.0, 128.0, 128.0, 0.0)) {
            Ok(m) => m,
            Err(_) => return vec![],
        };
        let roi = opencv::core::Rect::new(pad_x, pad_y, new_w, new_h);
        if let Ok(mut target_roi) = letterbox.roi_mut(roi) {
            if resized.copy_to(&mut target_roi).is_err() { return vec![]; }
        } else {
            return vec![];
        }
        drop(resized);

        // Convert BGR letterbox → CHW float [0, 1] (fast bulk operation)
        let ts_u = ts as usize;
        let total = (ts_u * ts_u) as usize;
        let mut float_data = vec![0.0f32; total * 3];

        if let Ok(pixels) = letterbox.data_bytes() {
            for i in 0..total {
                let b = pixels[i * 3] as f32 / 255.0;
                let g = pixels[i * 3 + 1] as f32 / 255.0;
                let r = pixels[i * 3 + 2] as f32 / 255.0;
                float_data[i] = b;
                float_data[total + i] = g;
                float_data[2 * total + i] = r;
            }
        }

        // Run ncnn inference
        let ex = self.net.create_extractor();
        if ex.input_bgr_normalized("in0", &float_data, ts, ts, 3).is_err() {
            return vec![];
        }
        let out_mat = match ex.extract("out0") {
            Ok(m) => m,
            Err(_) => return vec![],
        };

        let data = out_mat.data_f32();
        let (w, h, _c) = out_mat.shape();

        // ncnn Mat shape: (w, h, c) = (spatial, channels, batch)
        let num_dets = w as usize;
        let num_channels = h as usize;

        if num_channels < 5 || data.len() < num_channels * num_dets {
            return vec![];
        }

        let num_classes = num_channels - 4; // 4 bbox + class scores

        let mut candidates: Vec<Detection> = Vec::new();

        let mut raw_count = 0usize;
        for i in 0..num_dets {
            let cx = data[i];
            let cy = data[1 * num_dets + i];
            let w = data[2 * num_dets + i];
            let h = data[3 * num_dets + i];

            let mut best_conf = 0.0f32;
            let mut best_cls = 0i32;
            for c in 0..num_classes {
                let conf = data[(4 + c) * num_dets + i];
                if conf > best_conf { best_conf = conf; best_cls = c as i32; }
            }

            if best_conf < self.conf_thresh { continue; }
            raw_count += 1;
            if !enabled_classes.is_empty() && !enabled_classes.contains(&best_cls) { continue; }

            // Map from padded letterbox space to original frame
            let x1 = ((cx - w / 2.0 - pad_x as f32) / scale) as i32;
            let y1 = ((cy - h / 2.0 - pad_y as f32) / scale) as i32;
            let x2 = ((cx + w / 2.0 - pad_x as f32) / scale) as i32;
            let y2 = ((cy + h / 2.0 - pad_y as f32) / scale) as i32;

            let x1 = x1.max(0).min(fw - 1);
            let y1 = y1.max(0).min(fh - 1);
            let x2 = x2.max(0).min(fw - 1);
            let y2 = y2.max(0).min(fh - 1);

            if x2 <= x1 || y2 <= y1 { continue; }
            candidates.push(Detection {
                bbox: [x1, y1, x2, y2],
                centroid: ((x1 + x2) / 2, (y1 + y2) / 2),
                confidence: best_conf,
                class_id: best_cls,
                label: COCO_CLASSES.get(best_cls as usize).unwrap_or(&"?").to_string(),
            });
        }

        candidates
    }
}

fn resolve_path(path: &str) -> String {
    if std::path::Path::new(path).exists() {
        return path.to_string();
    }
    // Try relative to cwd/models/
    let alt = format!("models/{}", path.trim_start_matches("../models/"));
    if std::path::Path::new(&alt).exists() {
        return alt;
    }
    path.to_string()
}

fn non_max_suppression(mut dets: Vec<Detection>, iou_thresh: f32) -> Vec<Detection> {
    if dets.is_empty() { return dets; }
    dets.sort_by(|a, b| b.confidence.partial_cmp(&a.confidence).unwrap_or(std::cmp::Ordering::Equal));
    let mut keep = Vec::new();
    while let Some(best) = dets.pop() {
        keep.push(best.clone());
        dets.retain(|d| {
            let iou = iou(&d.bbox, &best.bbox);
            iou < iou_thresh
        });
    }
    keep
}

fn iou(a: &[i32; 4], b: &[i32; 4]) -> f32 {
    let x1 = a[0].max(b[0]);
    let y1 = a[1].max(b[1]);
    let x2 = a[2].min(b[2]);
    let y2 = a[3].min(b[3]);
    let inter = (x2 - x1).max(0) * (y2 - y1).max(0);
    let area_a = (a[2] - a[0]) * (a[3] - a[1]);
    let area_b = (b[2] - b[0]) * (b[3] - b[1]);
    inter as f32 / (area_a + area_b - inter) as f32 + 1e-6
}
